linera-views/src/reentrant_collection_view.rs

```diff
@@ -178,48 +178,83 @@ where
         let mut view = W::load(context).await?;
-        if clear {
+        if delete_storage_first {
             view.clear();
         }
         Ok(Arc::new(RwLock::new(view)))
     }
 
-    async fn try_load_view(
+    /// Reads the view if existing, otherwise report an error.
+    async fn wrapped_view_check(
         context: &C,
-        updates: &mut BTreeMap<Vec<u8>, Update<Arc<RwLock<W>>>>,
-        was_cleared: bool,
+        delete_storage_first: bool,
         short_key: &[u8],
-    ) -> Result<Arc<RwLock<W>>, ViewError> {
-        use btree_map::Entry::*;
+    ) -> Result<Option<Arc<RwLock<W>>>, ViewError> {
+        let key_index = context.base_tag_index(KeyTag::Index as u8, short_key);
+        if delete_storage_first || !context.contains_key(&key_index).await? {
+            Ok(None)
+        } else {
+            Ok(Some(
+                Self::wrapped_view(context, delete_storage_first, short_key).await?,
+            ))
+        }
+    }
 
+    /// Load the view and insert it into the updates if needed.
+    /// If the entry is missing, then it is set to default.
+    async fn try_load_view_mut(&mut self, short_key: &[u8]) -> Result<Arc<RwLock<W>>, ViewError> {
+        use btree_map::Entry::*;
+        let updates = self.updates.get_mut();
         Ok(match updates.entry(short_key.to_owned()) {
             Occupied(mut entry) => match entry.get_mut() {
                 Update::Set(view) => view.clone(),
                 entry @ Update::Removed => {
-                    let wrapped_view = Self::wrapped_view(context, true, short_key).await?;
+                    let wrapped_view = Self::wrapped_view(&self.context, true, short_key).await?;
                     *entry = Update::Set(wrapped_view.clone());
                     wrapped_view
                 }
             },
             Vacant(entry) => {
-                let wrapped_view = Self::wrapped_view(context, was_cleared, short_key).await?;
+                let wrapped_view =
+                    Self::wrapped_view(&self.context, self.delete_storage_first, short_key).await?;
                 entry.insert(Update::Set(wrapped_view.clone()));
                 wrapped_view
             }
         })
     }
 
+    /// Load the view from the update is available.
+    /// If missing, then the entry is loaded from storage and if
+    /// missing there an error is reported.
+    async fn try_load_view(&self, short_key: &[u8]) -> Result<Option<Arc<RwLock<W>>>, ViewError> {
+        let updates = self.updates.lock().await;
+        Ok(match updates.get(short_key) {
+            Some(entry) => match entry {
+                Update::Set(view) => Some(view.clone()),
+                _entry @ Update::Removed => {
+                    Self::wrapped_view_check(&self.context, true, short_key).await?
+                }
+            },
+            None => {
+                Self::wrapped_view_check(&self.context, self.delete_storage_first, short_key)
```

Comment:
`if self.delete_storage_first { None } else { ... }` ?

Reply:
But we also need to test for the existence of the key. So, I am afraid, it is not that simple.